lib/xgu/xgu.h

```diff
@@ -0,0 +1,535 @@
+inline
+DWORD* xgu_set_vertex_data_array_offset(DWORD* p, XguVertexArray index, const uint32_t data) {
+    return push_command_parameter(p, NV097_SET_VERTEX_DATA_ARRAY_OFFSET + index*4,
+                                  data/* & 0x03fffff*/);
+}
+
+inline
+DWORD* xgu_elements16(DWORD* p, uint16_t* elements, uint32_t count) {
+    assert(count % 2 == 0);
+    // This one is a bit complicated
+    assert(false);
+    return p;
+}
+
+inline
+DWORD* xgu_elements32(DWORD* p, uint32_t* elements, uint32_t count) {
+    // This one is a bit complicated
+    assert(false);
+    return p;
+}
+
+inline
+DWORD* xgu_draw_arrays(DWORD* p, uint32_t start, uint32_t count) {
+    return push_command_parameter(p, 0x40000000|NV097_DRAW_ARRAYS,
+                                  XGU_MASK(NV097_DRAW_ARRAYS_COUNT, (count-1)) |
+                                  XGU_MASK(NV097_DRAW_ARRAYS_START_INDEX, start));
+}
+
+/* ==== Alpha/Blend/Cull ==== */
+inline
+DWORD* xgu_set_alpha_test_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_ALPHA_TEST_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_blend_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_BLEND_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_cull_face_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_CULL_FACE_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_depth_test_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_DEPTH_TEST_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_dither_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_DITHER_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_lighting_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_LIGHTING_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_stencil_test_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_STENCIL_TEST_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_alpha_func(DWORD* p, uint32_t func) {
+    return push_command_parameter(p, NV097_SET_ALPHA_FUNC, func);
+}
+
+inline
+DWORD* xgu_set_alpha_ref(DWORD* p, uint32_t ref) {
+    return push_command_parameter(p, NV097_SET_ALPHA_REF, ref);
+}
+
+inline
+DWORD* xgu_set_blend_func_sfactor(DWORD* p, XguBlendFactor sf) {
+    return push_command_parameter(p, NV097_SET_BLEND_FUNC_SFACTOR, sf);
+}
+
+inline
+DWORD* xgu_set_blend_func_dfactor(DWORD* p, XguBlendFactor df) {
+    return push_command_parameter(p, NV097_SET_BLEND_FUNC_DFACTOR, df);
+}
+
+inline
+DWORD* xgu_set_color_mask(DWORD* p, XguColorMask cm) {
+    return push_command_parameter(p, NV097_SET_COLOR_MASK, cm);
+}
+
+inline
+DWORD* xgu_set_front_polygon_mode(DWORD* p, XguPolygonMode pm) {
+    return push_command_parameter(p, NV097_SET_FRONT_POLYGON_MODE, pm);
+}
+
+inline
+DWORD* xgu_set_cull_face(DWORD* p, XguCullFace cf) {
+    return push_command_parameter(p, NV097_SET_CULL_FACE, cf);
+}
+
+inline
+DWORD* xgu_set_front_face(DWORD* p, bool ccw) {
```

Comment:
Consider `xgu_set_front_face_ccw` or using an enum `XGU_FRONT_FACE_CW` / `_CCW` so code using this will be more self-documenting.